### PR TITLE
Removed Assets since tests was broken

### PR DESCRIPTION
**Pull request by jonasfj**
I suggest we remove `Assets` because:

 * There are no examples
 * Documentation doesn't say if it serves `build/` or `build/web/`
 * Users can easily roll their own static file serving logic

IMO it would be nice if the `appengine` experience was less magic... I guess this is a small step in that direction too. I'm well aware that I'm still a good deal away from getting rid of all the zone scoped globals.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.6.0
+
+ * Removed poorly documented assets support with broken tests.
```
Comment (sigurdm) [inline CHANGELOG.md]: Mark this as 
Breaking change:

and mention the specific API that is removed.
Also to be really nice you could write a snippet showing how to roll your own

**Comment by enyo** [comment]
Sorry if I'm mistaken, but the `assets` functionality did a lot more than simply read a file from disk.

In my opinion it's a big mistake to remove this from the library, since it takes care of serving files from a development pub server **or** an already generated file from disk.

The big advantage of using the `appengine` assets functionality is that I don't need to know whether I'm running in dev or production mode, and I don't need to implement any logic to either fetch the file from a server or from disk.

All the relevant code was in `lib/src/server/assets.dart`.

@mkustermann & @sigurdm would you please consider reverting this PR.

**Comment by mkustermann** [comment]
> would you please consider reverting this PR.

@jonasfj Is taking charge of the server side packages - so it is up to him to decide.

Personally I agree it's not good to remove (working) functionality without a better replacement.

/cc @mit-mit

**Comment by jonasfj** [comment]
@enyo, the tests were broken, and there was no documentation to suggest what it was serving.
Writing up something to serve static files should be pretty easy.. Let's make an example and if necessary add appengine specific support code if anything is needed.

I'm not oppose to having code that makes it easy to get started. But it should be documented, I couldn't fix tests because I had no idea what the correct logic was supposed to do.

Can you explain what this was supposed to do?

How is the logic any different depending on whether you're in development mode or production?

What folder was this supposed to be serving from? Relative to location of the script, `.packages` or current working directory?

**Comment by enyo** [comment]
@jonasfj it might not be the case anymore, because there are lots of changes since dart 2 (with the way files are built and served), but basically, when you start developing, you serve your files with `pub serve`. This launches an http server, that takes care of "compiling" the files, and serving them via http instead of simply building to the hard drive.
The assets library then had two modes: 1) production mode where it loaded the files from the HD, and 2) development mode, where it loaded the files from this development http server.

It might be, that nowadays, `pub serve` *always* builds the files to `build/` and that the http server is not technically necessary, in which case removing this functionality is fine. I have to look into that.

**Comment by jonasfj** [comment]
@enyo, do you mean that the assets library would load files from the development server in-order to ensure they were built? Would it also run `pub serve` or did you have run that in a separate terminal next to your app for this to work?

I think it would be cool to have this kind of development helper... but doesn't have to be appengine specific... though it would be neat to have example here pointing out how to use a community package for doing this.

**Comment by enyo** [comment]
@jonasfj the assets library would use the http server that `pub serve` (or `webdev serve`) launches to make sure the files are always the current version. Otherwise as a developer, you need to run `pub/webdev build` after each change.

From what I see, `webdev serve` supports an output directory though, so it could theoretically simply write the files to disk as well.

**Comment by enyo** [comment]
> @enyo, do you mean that the assets library would load files from the development server in-order to ensure they were built?

And yes, that's also a nice side effect: the temporary development http server makes sure the files are built before sending them.

**Comment by jonasfj** [comment]
> the assets library would use the http server that pub serve (or webdev serve) launches to make sure the files are always the current version......

The http handler or the http server? One implies proxying requests to another process on localhost.
I suspect you mean the http handler? (just curious).

But I see it ensures:
 A) file are rebuilt before being served (incoming request can be blocked until building is done), and,
 B) files don't have to proactively built before they are requested (if `webdev` supports this).

It seems to me that this makes sense for people who are using `webdev`. But since this is no longer built into `pub` I think it needs to live in a separate package. Something like this would need to depend on `webdev` and import the `webdev serve` logic if running in development mode.

Basically, you want something that returns an HTTP handler which does what `webdev serve` does... and is able to detect whether it's running in production or local development. I see this could be useful.

**Comment by jonasfj** [comment]
I filed https://github.com/dart-lang/webdev/issues/181 to expose the `webdev serve` logic...

I'm not 100% sure this is a good idea, maybe it's best to only use `webdev` as a `dev_dependency`.

Note. I'm not sure any of this worked post Dart 2.0, but I'm not 100% sure how the web development stuff works...

**Comment by enyo** [comment]
> The http handler or the http server? One implies proxying requests to another process on localhost.
I suspect you mean the http handler? (just curious).

I meant the http server. The previous assets implementation did an actual http request to the pub http server and returned the file, proxying it.

If the file `index.html` was requested, then it checked if `DART_PUB_SERVE` was set, and if it was (for example to `http://127.0.0.1:7888`) then it did an http request to `http://127.0.0.1:7888/index.html`, otherwise it served the file from disk: `./build/web/index.html`.

**Comment by jonasfj** [comment]
> I meant the http server. The previous assets implementation did an actual http request to the pub http server and returned the file, proxying it.

Okay, so you had to manually run `pub_serve` next to the appengine local service... right?

**Comment by enyo** [comment]
> Okay, so you had to manually run pub_serve next to the appengine local service... right?

Exactly. Or `webdev serve`. That's how the development environment also works atm. You write your angular/web app, use `webdev serve` (or previously `pub serve`) to start an interim http server that exposes all built files, and with appengine, you simply use the `assets` library to proxy the file from the interim http server.

**Comment by jonasfj** [comment]
Okay, I see no reason this logic should live in the `appengine` package, but I also don't necessarily think it's a bad idea -- given that it was already here...

If someone wants to add it back by:
1) Documenting and defining correct behavior,
2) Fix tests cases (such that `pub run test` passes),
3) Write an example of how to use this.

, then I would happy to do the review and land it :)

----
But I generally think this would be better located in a separate package. And I would be happy to recommend relevant community packages from the README in this package.
